### PR TITLE
fix(@libp2p/mdns): do not send TXT records that are too long

### LINKED ISSUE #2012: mDNS peer discovery stops working once a certain threshold is reached
mDNS peer discovery works by making a broadcast query for `_p2p._udp.local` services.  Peers respond with their multiaddrs.

The number of multiaddrs a peer has affects the size of the UDP message being sent as the mDNS response.

Once this message hits a certain size the `multicast-dns` module fails to decode the message.  The message size this occurs at is well under the maximum size for a mDNS response message.

Refs: https://github.com/mafintosh/multicast-dns/issues/85

### PR DESCRIPTION
**Pull request by achingbrain**
The data field of mDNS TXT records has a [hard limit of 255 characters](https://agnat.github.io/node_mdns/user_guide.html#txt_records) - some multiaddrs can be longer than this so filter them out before sending, otherwise remote peers can fail to parse our mDNS response.

We should also only be sending [link-local addresses](https://github.com/libp2p/specs/blob/master/discovery/mdns.md#issues) in mDNS responses so filter any non-link-local addresses out too, though go-libp2p includes loopback addresses even though it probably shouldn't (according to the mDNS spec which conflicts with our mDNS Peer Discovery spec) so we continue to do so as well.

Fixes #2012

### CODE REVIEW

File: packages/peer-discovery-mdns/src/index.ts [hunk 1]
```diff
@@ -113,14 +117,12 @@ class MulticastDNS extends EventEmitter<PeerDiscoveryEvents> implements PeerDisc
     }
   }
 
-  _onPeer (evt: CustomEvent<PeerInfo>): void {
-    if (this.mdns == null) {
-      return
-    }
+  _onMdnsWarning (err: Error): void {
+    log.error('mdns warning', err)
```
Comment (SgtPooki) [inline packages/peer-discovery-mdns/src/index.ts:121]: Should this be .error or .info?

File: packages/peer-discovery-mdns/src/query.ts [hunk 2]
```diff
@@ -92,20 +103,45 @@ export function gotQuery (qry: QueryPacket, mdns: MulticastDNS, peerName: string
...
+      // mDNS requires link-local addresses only
+      // https://github.com/libp2p/specs/blob/master/discovery/mdns.md#issues
+      .filter(isLinkLocal)
+      .forEach((addr) => {
+        const data = 'dnsaddr=' + addr.toString()
+
+        // TXT record fields have a max data length of 255 bytes
+        // see 6.1 - https://www.ietf.org/rfc/rfc6763.txt
+        if (data.length > 255) {
+          log('multiaddr %a is too long to use in mDNS query response', addr)
```
Comment (SgtPooki) [inline packages/peer-discovery-mdns/src/query.ts:116]: This will likely lead to a lot of circuit relay webtransport multiaddrs failing, won't it?

File: packages/peer-discovery-mdns/test/multicast-dns.spec.ts [hunk 3]
```diff
@@ -141,15 +142,6 @@ describe('MulticastDNS', () => {
     await stop(mdnsC)
   })
 
-  it('should start and stop with go-libp2p-mdns compat', async () => {
```
Comment (SgtPooki) [inline packages/peer-discovery-mdns/test/multicast-dns.spec.ts:144]: Why remove this test?

File: packages/peer-discovery-mdns/test/multicast-dns.spec.ts [hunk 4]
```diff
@@ -219,4 +211,80 @@ describe('MulticastDNS', () => {
...
+
+    const { detail: { id, multiaddrs } } = await new Promise<CustomEvent<PeerInfo>>((resolve) => {
+      mdnsA.addEventListener('peer', resolve, {
+        once: true
+      })
+    })
+
+    expect(pB.toString()).to.eql(id.toString())
+
+    ;[
```
Comment (SgtPooki) [inline packages/peer-discovery-mdns/test/multicast-dns.spec.ts:267]: Pre-colon? 🤢

**Review by SgtPooki — APPROVED** [review]
The tests are hard to follow but the src code changes make sense.

File: packages/peer-discovery-mdns/test/multicast-dns.spec.ts (hunk 3 above)
Comment (achingbrain) [reply packages/peer-discovery-mdns/test/multicast-dns.spec.ts:144]: Historically go-libp2p-mdns didn't speak spec-compliant mDNS so we had a compatibility layer.

go-libp2p-mdns has since been updated to behave better so the compatibility layer was removed, this test should have gone too.
